docs/design/proposals/VRG-Reconciliation.md

```diff
@@ -0,0 +1,132 @@
+kind: VolumeReplicationGroup
+metadata:
+  name: vrg-sample
+spec:
+  pvcSelector:
+    matchLabels:
+      appname: myapp
+  volumeReplicationClass: "vr-class-sample"
+  replicationState: "Primary"
+  s3Endpoint: http://objectstore.svc:9000
+  s3SecretName: vrg-objectstore-secret
+```
+
+## Reconciling VRG as Primary
+
+1. Select all PVCs that match `PVCSelector` for processing
+1. For each PVC
+    1. If not bound, continue but requeue
+    1. Check `vr-protected` finalizer, if found,
+        1. Check and remove (stale) `vr-protecting` finalizer
+        1. Check and update VR as `primary` and continue
+    1. Add `vr-protecting` finalizer to PVC, to prevent PVC deletion till VRG,
+    and hence VR is deleted
+    1. Change PV `reclaimPolicy` to "Retain", if not already "Retain" and add
+    an annotation to remember the change
+    1. [Find|Create] VR for PVC as `primary`
+    1. Backup PV bound to PVC
+    1. Add `vr-protected` finalizer to PVC
+    1. Remove `vr-protecting` finalizer
+1. For each PVC
+    1. Check if VR status has reached Primary state
+        1. If not, update status as progressing
+    1. Report status as completed, if not progressing for any PVC
+
+### Notes
+
+- `vr-protecting` and `vr-protected` are provided as 2 finalizers, to aid in
+  optimizing reconciliation if a PVC is already `vr-protected`
+
+## Reconciling VRG as Secondary
+
+1. Select all PVCs that match `PVCSelector` for processing
+1. For each PVC
+    1. If not bound, continue without requeue
+        1. **NOTE** We need PVC to be in deleted state, to prevent use by pods
+        when marking it as secondary. If it is still not bound and is deleted,
+        it will never be bound. If not deleted we cannot mark it secondary
+        anyway. So we skip without requeuing for this PVC.
+    1. Check `vr-protected` finalizer,
+        1. If found, check and remove (stale) `vr-protecting` finalizer
+        1. If not found, continue no requeue
+            1. **NOTE** If being deleted then we cannot add our finalizers as
+            we enable it to be replication protected, if not being deleted
+            then we cannot ensure use by pods as we setup volume as secondary.
+            Hence, we skip without requeuing for this PVC.
+    1. Ensure PVC being deleted; to prevent inadvertent use by pods as we
+    change state to Secondary
+        1. If not being deleted, Continue to next PVC, but requeue VRG
+        1. **NOTE** The pods and the PVCs are expected to be not in use when
+        marking VR as Secondary, and hence are expected to be deleted before
+        we take action on the same
+        1. **NOTE** With OCM as the control plane, the VRG would be marked
+        `seconday` and then deleted, ensuring we get to a clean `secondary`
+        state before deleting the VRs for each PVC
+    1. Ensure PVC not [in use](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#storage-object-in-use-protection)
```

Comment:
The operator is protecting the VR based on the VRG. User deletions may need finalizers, but a user should never really touch the VRs that we create as these are managed by us.

Barring external deletions (like the user) I do not see a need for VR deletion protection. BUT, adding it wont hurt, so that we get to process the VR even on such cases.

(keeping this comment open for addition to the doc.)